packages/core/src/utils/rpc.ts

```diff
@@ -147,59 +147,144 @@ export const _eth_getTransactionReceipt = (
-            : undefined,
-          toAddress: params.toAddress
-            ? params.toAddress.map((a) => toLowerCase(a))
-            : undefined,
-        },
+        typeof blockNumber === "number"
+          ? numberToHex(blockNumber)
+          : blockNumber,
+        { tracer: "callTracer" },
       ],
-    } as any)
-    .then((traces) => traces as unknown as SyncTrace[]);
+    })
+    .then((traces) => {
+      const result: SyncTrace[] = [];
+      let index = 0;
+      // all traces that weren't included because the trace has an error
+      // or the trace's parent has an error, mapped to the error string
+      const failedTraces = new Map<
+        (typeof traces)[number]["result"],
+        { error?: string; revertReason?: string }
+      >();
+
+      const dfs = (
+        frames: (typeof traces)[number]["result"][],
+        transactionHash: Hex,
+        parentFrame: (typeof traces)[number]["result"] | undefined,
+      ) => {
+        for (const frame of frames) {
+          if (frame.error !== undefined) {
+            failedTraces.set(frame, {
+              error: frame.error,
+              revertReason: frame.revertReason,
+            });
+          } else if (parentFrame && failedTraces.has(parentFrame)) {
+            const error = failedTraces.get(parentFrame)!;
+
+            frame.error = error.error;
+            frame.revertReason = error.revertReason;
+
+            failedTraces.set(frame, error);
+          }
+
+          // @ts-ignore
+          frame.index = index;
+          // @ts-ignore
+          frame.subcalls = frame.calls?.length ?? 0;
+
+          result.push({ trace: frame as SyncTrace["trace"], transactionHash });
+
+          index++;
+
+          if (frame.calls) {
+            dfs(frame.calls, transactionHash, frame);
+          }
+        }
+      };
+
+      for (const trace of traces) {
+        index = 0;
+        dfs([trace.result], trace.txHash, undefined);
+      }
+
+      return result;
+    });
 
 /**
- * Helper function for "trace_block" request.
+ * Helper function for "debug_traceBlockByHash" request.
  */
-export const _trace_block = (
+export const _debug_traceBlockByHash = (
   requestQueue: RequestQueue,
-  params: {
-    blockNumber: Hex | number;
+  {
+    hash,
+  }: {
+    hash: Hash;
   },
 ): Promise<SyncTrace[]> =>
   requestQueue
     .request({
-      method: "trace_block",
-      params: [
-        typeof params.blockNumber === "number"
-          ? numberToHex(params.blockNumber)
-          : params.blockNumber,
-      ],
-    } as any)
-    .then((traces) => traces as unknown as SyncTrace[]);
+      method: "debug_traceBlockByHash",
+      params: [hash, { tracer: "callTracer" }],
+    })
+    .then((traces) => {
+      const result: SyncTrace[] = [];
+      let index = 0;
+      // all traces that weren't included because the trace has an error
+      // or the trace's parent has an error, mapped to the error string
+      const failedTraces = new Map<
```

Comment:
Seems fine but I think this Map will contain frames across multiple transactions within the block